dev/io.openliberty.microprofile.telemetry.2.0.internal/src/io/openliberty/microprofile/telemetry20/internal/http/MPTelemetryHTTPMetricsAdapterImpl.java

```diff
@@ -82,19 +85,36 @@ public void updateHttpMetrics(HttpStatAttributes httpStatAttributes, Duration du
          * The Meter is built using the same static values each time.
          * The instrument that is recorded/updated is distinct for each
          * http-route/response/method combination (corresponds with resolved attributes).
-         * 
+         *
+         * However if nothing prompts OTel to build a Meter metrics exporters will have nowhere to export to.
+         *
+         * For most TCK tests `@Inject Meter` is all that's needed to give OTel that poke, but for HTTPHistogramTest
+         * we need to build a meter at least once within the context of an app. Using TheadContextClassLoaders will
+         * allow the test to pass and avoid the performance hit of calling this too often.
+         *
+         * TODO: Is it a bug in open telemetry that meter exporters don't do anything if nothing builds a meter?
+         * Is it a bug in HTTPHistogramTest that it doesn't create its own?
+         *
+         * TODO: Work with the performence team to optimise this as it runs every request. Is ApplicationMetaData better?
```

Comment:
I'm a bit confused by the statement "if nothing prompts OTel to build a Meter metrics exporters will have nowhere to export to".

In the original code, you're calling `otelInstance.getMeterProvider().get(INSTR_SCOPE)` which builds a `Meter`.

The only thing I can see being different in the new code is that you now use a different `httpHistogram` for each different TCCL that you see.

Earlier in the method, you do `OpenTelemetryAccessor.getOpenTelemetryInfo().getOpenTelemetry();` and IIRC that will get the `OpenTelemetry` instance for the current application.

Is the main change here that you're getting a new `httpHistogram` for each new `OpenTelemetry` instance? (because the TCCL and current application change at the same time, so new TCCL means possibly a new `OpenTelemetry`)

Reply:
I also note that the code above is checking for `otelInstance == null`, but I didn't think `OpenTelemetryInfo.getOpenTelemetry()` ever returns `null` - it would return a no-op instance.

Reply:
I don't understand this code nearly well as I would like, but its actually the side effects of calling `build()` rather than which histogram we hold onto. The TCCL map is there to ensure we don't build this again and again, I am told that has a horrible performance cost.

The side effect we care about:

```	PeriodicMetricReader.register(CollectionRegistration) line: 120	
	SdkMeterProvider.<init>(List<RegisteredView>, IdentityHashMap<MetricReader,CardinalityLimitSelector>, List<MetricProducer>, Clock, Resource, ExemplarFilter, ScopeConfigurator<MeterConfig>) line: 98	
	SdkMeterProviderBuilder.build() line: 202```

Then if you look at https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/export/PeriodicMetricReader.java

`register()` replaces a no op CollectionRegistration with one that returns metrics. Which is read in the scheduled `doRun()`.

And I know it has to be run inside the context of the app, or at least, I know calling it inside AbstractOpenTelemetryInfoFactory's application starting code wasn't enough. What I don't know is why it has to be run at this specific time and place.

Reply:
I don't understand this code as much as I will like, but its a side effect of creating that histogram that gets everything working. https://github.com/OpenLiberty/open-liberty/issues/29560 has the stacks and details.

Reply:
For future reference, this was our eventual analysis of the problem:

> without this change, `MPTelemetryHTTPMetricsAdapterImpl` is a singleton which has a single `DoubleHistogram` for recording HTTP metrics.
> This is ok when OTel is used at the system level, there's one `OpenTelemetry` instance and so we can use the same histogram instance all the time.
> However, when OTel is used at the app level, there's one `OpenTelemetry` instance per app, and so there should be one histogram per app and the way the code is written, the first app that records http metrics will end up recording all the http metrics for the whole server
> Additionally, if the first app that records http metrics is one that we ignore (like one of the built in endpoints, or the liberty-arquillian user feature), then all the http metrics end up in the no-op OTel instance for that "app" which throws them away
>
> a reliable way to hit the defect would be:
> - enable OTel at app level for one app
> - ensure the first request does not go to that app

We also noted
> so if we have 2 otel sdk instances, and one has the histogram, I think that would mean that the metric exporter for the one with the histogram would have the content for all of the http endpoints, regardless of app.  So that would mean the downstream consumers would still get the data, but it would be attributed to the wrong resource/service.name
>
> But if the first http hit is to something like `/openapi` or `/health`, and the server is running with otel associated to the apps, then all http histogram metrics updates would go to the noop otel sdk associated with the runtime

Our solution (as reflected in the final code) was to cache the `DoubleHistogram`, using the `OpenTelemetry` instance as the key to the cache.

Reply:
I see you used a WeakHashMap.  I assume it is ok to create it more than once potentially with the race condition in non synchronized HashMaps?

Reply:
That's probably wrong. I don't think creating two instances is a problem but the docs say

> Non-concurrent implementations should override this method and, on a best-effort basis, throw a ConcurrentModificationException if it is detected that the mapping function modifies this map during computation.

I'm not sure if `WeakHashMap` does that, but we wouldn't want to throw a `ConcurrentModificationException` here, so we should probably synchronize.

I'm not even sure if calling `get()` without synchronizing is safe if another thread could write to the map at the same time so we probably need to synchronize on every access :(

Reply:
In general HashMaps now handle the concurrency by having get possibly return a miss even if the entry exists.  So if really depends on if you are ok with that.  Otherwise you can use a ConcurrentHashMap with WeakReference keys like is done in other places.  I have thought of making it a general utility in kernel utilities, but haven't done that yet. https://raw.githubusercontent.com/OpenLiberty/open-liberty/integration/dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/rest/RestRouteCache.java is an example where we do it.